cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ProjectWatcher.java

```diff
@@ -0,0 +1,125 @@
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.HashMap;
+import java.util.Map;
+
+import static java.nio.file.StandardWatchEventKinds.ENTRY_CREATE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_DELETE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+import static java.nio.file.StandardWatchEventKinds.OVERFLOW;
+
+/**
+ * Represents a watcher that looks into project file changes.
+ *
+ * @since 2201.10.0
+ */
+public class ProjectWatcher {
+    private final WatchService fileWatcher;
+    private final Map<WatchKey, Path> watchKeys;
+    private final RunCommand runCommand;
+
+    public ProjectWatcher(Path projectPath, RunCommand runCommand) throws IOException {
+        this.fileWatcher = FileSystems.getDefault().newWatchService();
+        this.runCommand = runCommand;
+        this.watchKeys = new HashMap<>();
+        registerFileTree(projectPath);
+    }
+
+    void watch() throws IOException, InterruptedException {
+        RunCommandExecutor thread = new RunCommandExecutor(runCommand);
+        thread.start();
+        while (true) {
+            WatchKey key;
+            key = fileWatcher.take();
+            Path dir = watchKeys.get(key);
+            if (dir == null) {
+                continue;
+            }
+            for (WatchEvent<?> event : key.pollEvents()) {
+                WatchEvent.Kind<?> kind = event.kind();
+                if (kind == OVERFLOW) {
+                    continue;
+                }
+                WatchEvent<Path> pathWatchEvent = cast(event);
+                Path changedFileName = pathWatchEvent.context();
+                Path changedFilePath = dir.resolve(changedFileName);
+                if (isValidFileChange(changedFilePath)) {
+                    thread.terminate();
+                    thread.join();
+                    thread = new RunCommandExecutor(runCommand);
+                    thread.start();
+                }
+                if (kind == ENTRY_CREATE && Files.isDirectory(changedFilePath)) {
+                    registerFileTree(changedFilePath);
+                }
+            }
+            boolean valid = key.reset();
+            if (!valid) {
+                watchKeys.remove(key);
+                if (watchKeys.isEmpty()) {
+                    break;
+                }
+            }
+        }
+    }
+
+    private boolean isValidFileChange(Path path) {
+        if (Files.isDirectory(path)) {
+            return false;
+        }
+        Path fileNamePath = path.getFileName();
+        if (fileNamePath == null) {
+            return false;
+        }
+        String fileName = fileNamePath.toString();
+        return fileName.endsWith(".bal") || fileName.endsWith(".toml") && !fileName.equals("Dependencies.toml");
```

Comment:
If a file in the resource directory got chanaged, don't we need to consider that? wdyt?